experimental/attestation_common/src/crypto.rs

```diff
@@ -0,0 +1,123 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::anyhow;
+use ring::{
+    aead::{self, BoundKey},
+    agreement, rand,
+};
+
+// `ring::aead` uses 96-bit (12-byte) nonces.
+// https://briansmith.org/rustdoc/ring/aead/constant.NONCE_LEN.html
+//
+// TODO(#2087): Use a unique nonce for each message.
+const NONCE: [u8; aead::NONCE_LEN] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+
+struct OneNonceSequence(Option<aead::Nonce>);
+
+impl OneNonceSequence {
+    fn new() -> Self {
+        let nonce = aead::Nonce::assume_unique_for_key(NONCE);
+        Self(Some(nonce))
+    }
+}
+
+impl aead::NonceSequence for OneNonceSequence {
+    fn advance(&mut self) -> Result<aead::Nonce, ring::error::Unspecified> {
+        self.0.take().ok_or(ring::error::Unspecified)
+    }
+}
+
+pub struct AeadEncryptor {
+    // Key used for encrypting data.
+    sealing_key: aead::SealingKey<OneNonceSequence>,
+    // Key used for decrypting data.
+    opening_key: aead::OpeningKey<OneNonceSequence>,
+}
+
+impl AeadEncryptor {
+    pub fn new(key: &[u8]) -> Self {
+        let unbound_sealing_key = aead::UnboundKey::new(&aead::AES_256_GCM, &key).unwrap();
+        let sealing_key = ring::aead::SealingKey::new(unbound_sealing_key, OneNonceSequence::new());
+
+        let unbound_opening_key = aead::UnboundKey::new(&aead::AES_256_GCM, &key).unwrap();
+        let opening_key = ring::aead::OpeningKey::new(unbound_opening_key, OneNonceSequence::new());
+
+        Self {
+            sealing_key,
+            opening_key,
+        }
+    }
+
+    pub fn encrypt(&mut self, data: &[u8]) -> anyhow::Result<Vec<u8>> {
+        let mut encrypted_data = data.to_vec();
+        self.sealing_key
+            .seal_in_place_append_tag(
+                // Additional authenticated data is not required.
```

Comment:
Perhaps also include why it is not required?

Reply:
Done